### PR TITLE
rgw: don't return skew time error in pre-signed url

### PR DESCRIPTION
**Pull request by Liuchang0812**
Fixes: http://tracker.ceph.com/issues/18828

Signed-off-by: liuchang0812 <liuchang0812@gmail.com>

### CODE REVIEW

**Comment by Liuchang0812** [comment]
@yehudasa do you mind taking a look, thanks

**Comment by Liuchang0812** [comment]
I have tested it in my radosgw env, I generated a pre-signed URL that its expiry is 90000, more than our skew_time(7min). : 
```
http://rgw-test.us-east-1.s3.amazonaws.com:8000/5m.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expires=9000&X-Amz-SignedHeaders=host&X-Amz-Signature=4abed2f1622cdb5d9528f4392e8315cbb998c5c84aca81b50f60467ed87abd03&X-Amz-Date=20170210T121602Z&X-Amz-Credential=0343ADSQIMU3FIGK4LLT%2F20170210%2Fus-east-1%2Fs3%2Faws4_request`
```

I could download this file by this URL as

```
➜  build git:(wip-18828) ✗ wget "http://rgw-test.us-east-1.s3.amazonaws.com:8000/5m.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expires=9000&X-Amz-SignedHeaders=host&X-Amz-Signature=4abed2f1622cdb5d9528f4392e8315cbb998c5c84aca81b50f60467ed87abd03&X-Amz-Date=20170210T121602Z&X-Amz-Credential=0343ADSQIMU3FIGK4LLT%2F20170210%2Fus-east-1%2Fs3%2Faws4_request" -O 1.txt
--2017-02-10 20:58:31--  http://rgw-test.us-east-1.s3.amazonaws.com:8000/5m.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Expires=9000&X-Amz-SignedHeaders=host&X-Amz-Signature=4abed2f1622cdb5d9528f4392e8315cbb998c5c84aca81b50f60467ed87abd03&X-Amz-Date=20170210T121602Z&X-Amz-Credential=0343ADSQIMU3FIGK4LLT%2F20170210%2Fus-east-1%2Fs3%2Faws4_request
Resolving rgw-test.us-east-1.s3.amazonaws.com (rgw-test.us-east-1.s3.amazonaws.com)... 127.0.0.1
Connecting to rgw-test.us-east-1.s3.amazonaws.com (rgw-test.us-east-1.s3.amazonaws.com)|127.0.0.1|:8000... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5242880 (5.0M) [application/octet-stream]
Saving to: ‘1.txt’

1.txt                                              100%[================================================================================================================>]   5.00M  --.-KB/s    in 0.01s   

2017-02-10 20:58:32 (395 MB/s) - ‘1.txt’ saved [5242880/5242880]
```

And, some logs I added for debug:
```
2017-02-10 20:58:31.994389 7fb475067700 20 get authorization from request params 
2017-02-10 20:58:31.994536 7fb475067700 20 NOTICE: now = 1486731511, now_req = 1486728962, exp = 9000
2017-02-10 20:58:31.994543 7fb475067700 20 request time is not skew.             
```
so, `1486731511 - 1486728962 = 42min`, 42min > RGW_AUTH_GRACE_MINS(15min). it works.

**Comment by Liuchang0812** [comment]
Sorry for compiling error, I will fix it soon

**Comment by Liuchang0812** [comment]
I would like to add some unittest cases.  Which files is proper to add unittest? @yehudasa @mattbenjamin

**Comment by mattbenjamin** [comment]
@Liuchang0812 the right place would be in our s3-tests, I think

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -3587,12 +3591,12 @@ int RGW_Auth_S3::authorize_v4(RGWRados *store, struct req_state *s, bool force_b
       return -EPERM;
 
     s->aws4_auth->expires = s->info.args.get("X-Amz-Expires");
-    if (s->aws4_auth->expires.size() != 0) {
+    if (!s->aws4_auth->expires.empty()) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_s3.cc]: yeah, prefer this

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -3587,12 +3591,12 @@ int RGW_Auth_S3::authorize_v4(RGWRados *store, struct req_state *s, bool force_b
...
       /* X-Amz-Expires provides the time period, in seconds, for which
          the generated presigned URL is valid. The minimum value
          you can set is 1, and the maximum is 604800 (seven days) */
       time_t exp = atoll(s->aws4_auth->expires.c_str());
-      if ((exp < 1) || (exp > 604800)) {
+      if ((exp < 1) || (exp > 7*24*60*60)) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_s3.cc]: this is harmless, but not sure if it's preferable as the comment above elucidates the magic value

Comment (Liuchang0812) [inline src/rgw/rgw_rest_s3.cc]: need i reset this change?

**Comment by cbodley** [comment]
[teuthology run](http://pulpito.ceph.com/cbodley-2017-02-15_16:05:58-rgw-wip-cbodley-testing---basic-mira/) was clean (only failures were selinux related)

**Comment by Liuchang0812** [comment]
@mattbenjamin is it ok to be merged, please tell me if need do anything.

**Comment by Liuchang0812** [comment]
@mattbenjamin I have looked at s3-tests, but if we create a test-case in s3-tests, we must sleep more then 15 minutes and then visit the pre-sign url. is it ok? it makes test time too long.

**Comment by cbodley** [comment]
is there no way to give a custom Date header to boto to fake it?

**Comment by Liuchang0812** [comment]
@cbodley thanks for your suggestion. I will have a try.

**Comment by Liuchang0812** [comment]
@cbodley https://github.com/boto/boto/pull/3686

**Comment by cbodley** [comment]
@Liuchang0812 we'd like to merge this, even though the boto changes are pending. could you open a pull request against https://github.com/ceph/s3-tests that uses that fixed branch of boto?

**Comment by Liuchang0812** [comment]
@cbodley roger that

**Comment by Liuchang0812** [comment]
@cbodley could I add a git sub-module in s3-tests to use my boto branch? I tried to fix branch in `requirements.txt`, but it seems doesn't work.

**Comment by cbodley** [comment]
@Liuchang0812 i don't think we need to deal with the dependency for now. just push the s3tests part, and we can move forward once the boto change gets in

**Comment by Liuchang0812** [comment]
@cbodley I pushed a PR to s3-tests https://github.com/ceph/s3-tests/pull/155

**Comment by smithfarm** [comment]
Noting for posterity that http://tracker.ceph.com/issues/18829 claims that this PR fixes it, although the commit message says it fixes only http://tracker.ceph.com/issues/18828

Thus, the jewel backports of http://tracker.ceph.com/issues/18828 and http://tracker.ceph.com/issues/18829 were resolved by cherry-picking just this one commit. Hopefully that is correct.

**Pull request merged**